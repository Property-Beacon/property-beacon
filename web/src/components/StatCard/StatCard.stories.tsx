import { FiBarChart, FiRefreshCw } from 'react-icons/fi'
import StatCard from './StatCard'

export const variants = () => {
  return (
    <div className="flex">
      <StatCard title="Total submission" number={1000} className="mr-4">
        neutral (default)
      </StatCard>
      <StatCard
        title="Total submission"
        number={1000000}
        variant="primary"
        className="mr-4"
      >
        primary
      </StatCard>
      <StatCard
        title="Total submission"
        number={1000000000}
        variant="secondary"
      >
        secondary
      </StatCard>
      <StatCard
        title="Total submission"
        number={12.2}
        variant="accent"
        className="mr-4"
      >
        accent
      </StatCard>
      <StatCard
        title="Total submission"
        number={66.0}
        variant="info"
        className="mr-4"
      >
        info / 66.0
      </StatCard>
      <StatCard
        title="Total submission"
        number={345.678}
        variant="success"
        className="mr-4"
      >
        success
      </StatCard>
      <StatCard
        title="Total submission"
        number={123.1312321}
        variant="warning"
        className="mr-4"
      >
        warning / 123.1312321
      </StatCard>
      <StatCard title="Total submission" number={0.12345533364} variant="error">
        error / 0.12345533364
      </StatCard>
    </div>
  )
}

export const clickable = () => (
  <div className="flex">
    <StatCard
      to="#"
      variant="error"
      className="w-60"
      number={98765}
      title="Total submission"
    >
      This StatCard is clickable
    </StatCard>
  </div>
)

export const custom = () => {
  return (
    <div className="flex">
      <StatCard
        variant="success"
        title={
          <div className="flex place-items-center text-warning text-2xl">
            <FiBarChart />
            <span className="ml-1">Age group</span>
          </div>
        }
        number={
          <div className="text-5xl font-extrabold flex place-items-center mt-4 mb-8 text-primary">
            99999
            <button className="btn btn-accent btn-sm ml-2">
              <FiRefreshCw className="mr-2" />
              Refresh
            </button>
          </div>
        }
      >
        <div className="flex place-items-center">
          <div className="badge badge-outline">age range: 20~30</div>
        </div>
      </StatCard>
    </div>
  )
}

export default { title: 'Components/StatCard' }
