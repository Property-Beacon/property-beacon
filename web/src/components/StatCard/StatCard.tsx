import { Link } from '@redwoodjs/router'
import type { FC, HTMLAttributes, ReactNode } from 'react'

type BaseProps = {
  variant?: ColorVariant
  title: ReactNode
  number: ReactNode
  numberFormatOptions?: Intl.NumberFormatOptions
} & Omit<HTMLAttributes<HTMLDivElement>, 'title'>

const Base: FC<BaseProps> = ({
  variant = 'neutral',
  children,
  title,
  number,
  numberFormatOptions,
  className,
  ...rest
}) => {
  const numberFormatter = new Intl.NumberFormat(
    navigator.language,
    numberFormatOptions
  )
  const [textColor, borderColor] =
    variant === 'primary'
      ? ['text-primary', 'border-primary']
      : variant === 'secondary'
      ? ['text-secondary', 'border-secondary']
      : variant === 'accent'
      ? ['text-accent', 'border-accent']
      : variant === 'info'
      ? ['text-info', 'border-info']
      : variant === 'success'
      ? ['text-success', 'border-success']
      : variant === 'warning'
      ? ['text-warning', 'border-warning']
      : variant === 'error'
      ? ['text-error', 'border-error']
      : ['text-neutral', 'border-neutral']

  return (
    <div
      data-testid="stat-card"
      className={`bg-base-200 rounded-lg p-4 shadow-md relative overflow-hidden ${className}`}
      {...rest}
    >
      <div
        className={`absolute h-full border-l-4 left-0 top-0 ${borderColor}`}
      />
      {typeof title === 'string' ? (
        <div className="opacity-50" data-testid="stat-card-title">
          {title}
        </div>
      ) : (
        title
      )}
      {typeof number === 'bigint' || typeof number === 'number' ? (
        <div className="text-xl font-extrabold" data-testid="stat-card-number">
          {numberFormatter.format(number)}
        </div>
      ) : (
        number
      )}
      {typeof children === 'string' ? (
        <div
          data-testid="stat-card-children"
          className={`${textColor} text-xs`}
        >
          {children}
        </div>
      ) : (
        children
      )}
    </div>
  )
}

interface Props extends BaseProps {
  to?: string
}

const StatCard: FC<Props> = ({ to, className, ...rest }) => {
  return to ? (
    <Link to={to} className={className} data-testid="stat-card-link">
      <Base {...rest} className="hover:bg-base-300" />
    </Link>
  ) : (
    <Base {...rest} className={className} />
  )
}

export default StatCard
