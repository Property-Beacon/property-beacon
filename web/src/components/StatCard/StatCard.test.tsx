import { fireEvent, render, screen, waitFor } from '@redwoodjs/testing'
import { FiBarChart, FiRefreshCw } from 'react-icons/fi'
import StatCard from './StatCard'

describe('StatCard', () => {
  it('renders successfully', () => {
    expect(() => {
      render(
        <StatCard title="Total" number={999}>
          description
        </StatCard>
      )
    }).not.toThrow()

    expect(() => {
      screen.getByTestId('stat-card-link')
    }).toThrow()

    expect(screen.getByTestId('stat-card')).toMatchSnapshot()
    expect(screen.getByTestId('stat-card-title')).toHaveTextContent('Total')
    expect(screen.getByTestId('stat-card-number')).toHaveTextContent('999')
    expect(screen.getByTestId('stat-card-children')).toHaveTextContent(
      'description'
    )
  })

  it('renders clickable StatCard', async () => {
    const mockFn = jest.fn()

    render(
      <StatCard title="Count" number={99} to="#">
        This is clickable
      </StatCard>
    )

    const StatCardLink = screen.getByTestId('stat-card-link')

    expect(StatCardLink).toBeInTheDocument()
    expect(StatCardLink).toMatchSnapshot()
    expect(screen.getByTestId('stat-card-title')).toHaveTextContent('Count')
    expect(screen.getByTestId('stat-card-number')).toHaveTextContent('99')
    expect(screen.getByTestId('stat-card-children')).toHaveTextContent(
      'This is clickable'
    )

    StatCardLink.addEventListener('click', mockFn)

    fireEvent.click(screen.getByTestId('stat-card-link'))
    await waitFor(() => expect(mockFn).toHaveBeenCalledTimes(1))

    StatCardLink.removeEventListener('click', mockFn)
  })

  it('renders StatCard with custom title, number and children', () => {
    render(
      <StatCard
        variant="success"
        title={
          <div className="flex place-items-center text-warning text-2xl">
            <FiBarChart />
            <span className="ml-1">Age group</span>
          </div>
        }
        number={
          <div className="text-5xl font-extrabold flex place-items-center mt-4 mb-8 text-primary">
            99999
            <button className="btn btn-accent btn-sm ml-2">
              <FiRefreshCw className="mr-2" />
              Refresh
            </button>
          </div>
        }
      >
        <div className="flex place-items-center">
          <div className="badge badge-outline">age range: 20~30</div>
        </div>
      </StatCard>
    )

    expect(() => {
      screen.getByTestId('stat-card-title')
    }).toThrow()
    expect(() => {
      screen.getByTestId('stat-card-number')
    }).toThrow()
    expect(() => {
      screen.getByTestId('stat-card-children')
    }).toThrow()

    expect(screen.getByText('Age group')).toBeInTheDocument()
    expect(screen.getByText('99999')).toBeInTheDocument()
    expect(screen.getByText('age range: 20~30')).toBeInTheDocument()
  })
})
