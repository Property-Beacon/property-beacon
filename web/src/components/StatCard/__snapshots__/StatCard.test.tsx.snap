// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatCard renders successfully 1`] = `
<div
  class="bg-base-200 rounded-lg p-4 shadow-md relative overflow-hidden undefined"
  data-testid="stat-card"
>
  <div
    class="absolute h-full border-l-4 left-0 top-0 border-neutral"
  />
  <div
    class="opacity-50"
    data-testid="stat-card-title"
  >
    Total
  </div>
  <div
    class="text-xl font-extrabold"
    data-testid="stat-card-number"
  >
    999
  </div>
  <div
    class="text-neutral text-sm"
    data-testid="stat-card-children"
  >
    description
  </div>
</div>
`;
